chichiwang/exercise-js
# Rework npm scripts

## Description
Linked to Issue: N/A
Add [npm-run-all](https://www.npmjs.com/package/npm-run-all) to the project and rework the npm scripts to:
* Utilize [run-s](https://github.com/mysticatea/npm-run-all/blob/HEAD/docs/run-s.md) command for brevity
* Make the available scripts more intuitive

## Changes
* Add devDependency: `npm-run-all`
* Modify the following scripts in `package.json`:
  * `build:dev`: Add a `prebuild:dev` to run the `clean` script
  * `build:prod`: Add a `prebuild:prod` to run the `clean` script
  * `clean`: Add a step to remove `stats.json` generated from the build steps
  * `analyze:dev`: `build:dev:analyze` renamed
  * `analyze:prod`: `build:prod:analyze` renamed
  * `dev`: `build:dev:run` renamed
  * `prod`: `build:prod:run` renamed
  * `start`: Build and run the production build without piping output through `pino-pretty`

## Steps to QA
* Pull down and switch to this branch
* Run `npm build:dev`
  * Ensure `stats.json` and the `dist/` dir are cleaned prior to asset builds
* Run `npm run dev`
  * Ensure assets are built correctly, server starts up, and verify in browser
* Run `npm build:prod`
  * Ensure `stats.json` and the `dist/` dir are cleaned prior to asset builds
* Run `npm run prod`
  * Ensure assets are built correctly, server starts up, and verify in browser
* Run `npm start`
  * Ensure assets are built correctly, server starts up, and verify in browser
  * Ensure server logs are not prettified
* Run `npm run analyze:dev` and ensure bundle analyzer runs correctly
  * Ensure analyzer is display un-minified dev assets
* Run `npm run analyze:prod` and ensure bundle analyzer runs correctly
  * Ensure analyzer is display minified prod assets
